src/memory/array.rs

```diff
@@ -0,0 +1,545 @@
+    /// Signed 32-bit integer
+    SignedInt32,
+    /// Half-precision floating point number
+    Half,
+    /// Single-precision floating point number
+    Float,
+}
+
+impl ArrayFormat {
+    /// Creates ArrayFormat from the CUDA Driver API enum
+    pub fn from_raw(raw: CUarray_format) -> Self {
+        match raw {
+            CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT8 => ArrayFormat::UnsignedInt8,
+            CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT16 => ArrayFormat::UnsignedInt16,
+            CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT32 => ArrayFormat::UnsignedInt32,
+            CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT8 => ArrayFormat::SignedInt8,
+            CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT16 => ArrayFormat::SignedInt16,
+            CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT32 => ArrayFormat::SignedInt32,
+            CUarray_format_enum::CU_AD_FORMAT_HALF => ArrayFormat::Half,
+            CUarray_format_enum::CU_AD_FORMAT_FLOAT => ArrayFormat::Float,
+        }
+    }
+
+    /// Converts ArrayFormat to the CUDA Driver API enum
+    pub fn to_raw(self) -> CUarray_format {
+        match self {
+            ArrayFormat::UnsignedInt8 => CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT8,
+            ArrayFormat::UnsignedInt16 => CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT16,
+            ArrayFormat::UnsignedInt32 => CUarray_format_enum::CU_AD_FORMAT_UNSIGNED_INT32,
+            ArrayFormat::SignedInt8 => CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT8,
+            ArrayFormat::SignedInt16 => CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT16,
+            ArrayFormat::SignedInt32 => CUarray_format_enum::CU_AD_FORMAT_SIGNED_INT32,
+            ArrayFormat::Half => CUarray_format_enum::CU_AD_FORMAT_HALF,
+            ArrayFormat::Float => CUarray_format_enum::CU_AD_FORMAT_FLOAT,
+        }
+    }
+}
+
+bitflags! {
+    /// Flags which modify the behavior of CUDA array creation.
+    #[derive(Default)]
+    pub struct ArrayObjectFlags: c_uint {
+        /// Enables creation of layered CUDA array.s When this flag is set, depth specifies the
+        /// number of layers, not the depth of a 3D array.
+        const LAYERED = cuda_sys::cuda::CUDA_ARRAY3D_LAYERED;
+
+        /// Enables surface references to be bound to the CUDA array.
+        const SURFACE_LDST = cuda_sys::cuda::CUDA_ARRAY3D_SURFACE_LDST;
+
+        /// Enables creation of cubemaps. If this flag is set, Width must be equal to Height, and
+        /// Depth must be six. If the `LAYERED` flag is also set, then Depth must be a multiple of
+        /// six.
+        const CUBEMAP = cuda_sys::cuda::CUDA_ARRAY3D_CUBEMAP;
+
+        /// Indicates that the CUDA array will be used for texture gather. Texture gather can only
+        /// be performed on 2D CUDA arrays.
+        const TEXTURE_GATHER = cuda_sys::cuda::CUDA_ARRAY3D_TEXTURE_GATHER;
+    }
+}
+
+impl ArrayObjectFlags {
+    /// Creates a default flags object with no flags set.
```

Comment:
I don't have a problem with adding builder functions for the various flags objects, but we should be consistent about having them for all bitflags types or none.

Reply:
Yeah, I think I'll remove them. They don't add much.